src/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,179 @@
+
+#include <iomanip>
+
+#include "software/geom/algorithms/contains.h"
+#include "software/logger/logger.h"
+
+PassGenerator::PassGenerator(const TbotsProto::PassingConfig& passing_config)
+    : optimizer_(optimizer_param_weights),
+      random_num_gen_(RNG_SEED),
+      passing_config_(passing_config)
+{
+}
+
+PassWithRating PassGenerator::getBestPass(const World& world,
+                                          const std::vector<RobotId>& robots_to_ignore)
+{
+    auto receiving_positions_map =
+        sampleReceivingPositionsPerRobot(world, robots_to_ignore);
+
+    // if there are no friendly robots, return early
+    if (receiving_positions_map.empty())
+    {
+        // default pass with 0 rating
+        return PassWithRating{Pass(Point(), Point(), 1.0), 0};
+    }
+
+    // Optimize the receiving positions for each robot and get the best pass
+    PassWithRating best_pass = optimizeReceivingPositions(world, receiving_positions_map);
+
+    // Visualize the sampled passes and the best pass
+    if (passing_config_.pass_gen_vis_config().visualize_sampled_passes())
+    {
+        std::vector<TbotsProto::DebugShapes::DebugShape> debug_shapes;
+        for (const auto& [robot_id, receiving_positions] : receiving_positions_map)
+        {
+            for (const Point& receiving_position : receiving_positions)
+            {
+                debug_shapes.push_back(*createDebugShape(
+                    Stadium(world.friendlyTeam().getRobotById(robot_id)->position(),
+                            receiving_position, 0.02),
+                    std::to_string(debug_shapes.size()) + "pg"));
+            }
+        }
+        std::stringstream stream;
+        stream << "BP:" << std::fixed << std::setprecision(3) << best_pass.rating;
+        debug_shapes.push_back(
+            *createDebugShape(Circle(best_pass.pass.receiverPoint(), 0.05),
+                              std::to_string(debug_shapes.size()) + "pg", stream.str()));
+        LOG(VISUALIZE) << *createDebugShapes(debug_shapes);
+    }
+
+    // Generate sample passes across the field for cost visualization
+    if (passing_config_.cost_vis_config().generate_sample_passes())
+    {
+        samplePassesForVisualization(world, passing_config_, best_pass.pass);
+    }
+
+    return best_pass;
+}
+
+std::map<RobotId, std::vector<Point>> PassGenerator::sampleReceivingPositionsPerRobot(
+    const World& world, const std::vector<RobotId>& robots_to_ignore)
+{
+    std::map<RobotId, std::vector<Point>> receiving_positions_map;
+
+    const double min_sampling_std_dev =
+        passing_config_.pass_gen_min_rand_sample_std_dev_meters();
+    const double sampling_std_dev_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_std_dev_robot_vel_multiplier();
+    const double sampling_center_vel_multiplier =
+        passing_config_.pass_gen_rand_sample_center_robot_vel_multiplier();
+
+    for (const Robot& robot : world.friendlyTeam().getAllRobots())
```

Comment:
Do we really need to sample all the robots that are not either a goalie or has the ball?

For example, it is probably not a good idea to pass to robots way to close to an enemy robot?

Reply:
The idea here was that we know for sure that the we don't want to pass to the goalie, or the robot with the ball. There might be a slight chance that we would want to pass to any other robots, and that should be determined by the "cost functions" used to rate the pass. For example, we have `ratePassEnemyRisk` for taking into account enemies close to friendlies and the chance of an interception. IMO it's more intuitive to have all of the more complex logic within the cost functions, but I can see an argument for possibly improving performance by adding other heuristics like the one you mentioned to filter out additional friendly robots.

Reply:
if we need to, we can do dumb "soft" filtering and passing candidate robots in here. Right now, it seems like PassGenerator is quite fast, so I don't think we need this optimization